File: wpilibj/src/test/java/org/wpilib/smartdashboard/Field2dTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.smartdashboard;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.wpilib.math.geometry.Pose2d;
import org.wpilib.math.geometry.Rotation2d;
import org.wpilib.math.kinematics.ChassisAccelerations;
import org.wpilib.math.kinematics.ChassisVelocities;
import org.wpilib.math.trajectory.HolonomicSample;
import org.wpilib.math.trajectory.HolonomicTrajectory;
import org.wpilib.tunable.MockTunableBackend;
import org.wpilib.tunable.TunableConfig;
import org.wpilib.tunable.TunableRegistry;
import org.wpilib.tunable.Tunables;

class Field2dTest {
  private MockTunableBackend m_mock;

  @BeforeEach
  void setUp() {
    m_mock = new MockTunableBackend();
    TunableRegistry.reset();
    TunableRegistry.registerBackend("", m_mock);
  }

  @AfterEach
  void tearDown() {
    TunableRegistry.reset();
  }

  @Test
  void dashboardEditsRobotPoseThroughTunable() {
    Field2d field = new Field2d();
    Tunables.publish("field", field);

    Pose2d pose = new Pose2d(1.0, 2.0, Rotation2d.fromDegrees(30.0));
    m_mock.setArray("/field/Robot", new Pose2d[] {pose});
    TunableRegistry.update();

    assertEquals(pose, field.getRobotPose());
  }

  @Test
  void dashboardEditsObjectPosesThroughTunable() {
    Field2d field = new Field2d();
    final FieldObject2d existing = field.getObject("target");
    Tunables.publish("field", field);

    Pose2d first = new Pose2d(1.0, 2.0, Rotation2d.fromDegrees(30.0));
    Pose2d second = new Pose2d(3.0, 4.0, Rotation2d.fromDegrees(60.0));

    m_mock.setArray("/field/target", new Pose2d[] {first, second});
    TunableRegistry.update();

    assertEquals(List.of(first, second), existing.getPoses());
  }

  @Test
  void dashboardEditsObjectCreatedAfterTunablePublish() {
    Field2d field = new Field2d();
    Tunables.publish("field", field);
    final FieldObject2d late = field.getObject("target");

    Pose2d pose = new Pose2d(1.0, 2.0, Rotation2d.fromDegrees(30.0));

    m_mock.setArray("/field/target", new Pose2d[] {pose});
    TunableRegistry.update();

    assertEquals(List.of(pose), late.getPoses());
  }

  @Test
  void fieldObjectPoseTunablesUseChangeNotifications() {
    Field2d field = new Field2d();
    Tunables.publish("field", field);

    var tunable = m_mock.getTunable("/field/Robot");

    assertEquals(TunableConfig.Polling.GET_ON_CHANGE, tunable.getConfig().getPolling());
    assertFalse(tunable.hasChanged());
  }

  @Test
  void localObjectMutationsNotifyTunableBackend() {
    Field2d field = new Field2d();
    final FieldObject2d object = field.getObject("target");
    Tunables.publish("field", field);

    var tunable = m_mock.getTunable("/field/target");
    Pose2d first = new Pose2d(1.0, 2.0, Rotation2d.fromDegrees(30.0));

    tunable.resetChanged();
    object.setPose(first);
    assertTrue(tunable.hasChanged());

    Pose2d second = new Pose2d(3.0, 4.0, Rotation2d.fromDegrees(60.0));

    tunable.resetChanged();
    object.setPoses(List.of(first, second));
    assertTrue(tunable.hasChanged());

    tunable.resetChanged();
    object.setPoses(first, second);
    assertTrue(tunable.hasChanged());

    tunable.resetChanged();
    object.setTrajectory(
        new HolonomicTrajectory(
            List.of(
                new HolonomicSample(
                    0.0, first, new ChassisVelocities(), new ChassisAccelerations()),
                new HolonomicSample(
                    1.0, second, new ChassisVelocities(), new ChassisAccelerations()))));
    assertTrue(tunable.hasChanged());
    assertEquals(List.of(first, second), object.getPoses());
  }

  @Test
  void closeRemovesPublishedTunables() {
    Field2d field = new Field2d();
    field.getObject("target");
    Tunables.publish("field", field);
    Tunables.publish("fieldAlias", field);

    Pose2d staleEdit = new Pose2d(1.0, 2.0, Rotation2d.fromDegrees(30.0));
    m_mock.setArray("/field/Robot", new Pose2d[] {staleEdit});
    field.close();
    TunableRegistry.update();

    assertEquals(Pose2d.ZERO, field.getRobotPose());

    assertThrows(
        IllegalArgumentException.class,
        () -> m_mock.setArray("/field/Robot", new Pose2d[] {staleEdit}));
    assertThrows(
        IllegalArgumentException.class,
        () -> m_mock.setArray("/field/target", new Pose2d[] {staleEdit}));
    assertThrows(
        IllegalArgumentException.class,
        () -> m_mock.setArray("/fieldAlias/Robot", new Pose2d[] {staleEdit}));

    Field2d replacement = new Field2d();
    Tunables.publish("field", replacement);
    Tunables.publish("fieldAlias", new Field2d());

    Pose2d replacementPose = new Pose2d(3.0, 4.0, Rotation2d.fromDegrees(60.0));
    m_mock.setArray("/field/Robot", new Pose2d[] {replacementPose});
    TunableRegistry.update();

    assertEquals(replacementPose, replacement.getRobotPose());
  }
}

File: wpilibc/src/test/native/cpp/smartdashboard/Field2dTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/smartdashboard/Field2d.hpp"

#include <memory>
#include <utility>
#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "wpi/tunables/MockTunableBackend.hpp"
#include "wpi/tunables/TunableRegistry.hpp"
#include "wpi/tunables/Tunables.hpp"
#include "wpi/units/angle.hpp"

struct Field2dTest {
  Field2dTest() {
    wpi::tunables::TunableRegistry::Reset();
    wpi::tunables::TunableRegistry::RegisterBackend("", mock);
  }

  ~Field2dTest() { wpi::tunables::TunableRegistry::Reset(); }

  std::shared_ptr<wpi::tunables::MockTunableBackend> mock =
      std::make_shared<wpi::tunables::MockTunableBackend>();
};

static void CheckPose(const wpi::math::Pose2d& actual,
                      const wpi::math::Pose2d& expected) {
  CHECK(actual.X() == expected.X());
  CHECK(actual.Y() == expected.Y());
  CHECK(actual.Rotation().Degrees() == expected.Rotation().Degrees());
}

TEST_CASE_METHOD(Field2dTest,
                 "Field2dTest DashboardEditsRobotPoseThroughTunable",
                 "[wpilibc][smartdashboard]") {
  wpi::Field2d field;
  wpi::tunables::Publish("field", field);

  wpi::math::Pose2d pose{1_m, 2_m, 30_deg};
  mock->SetStructVector<wpi::math::Pose2d>("/field/Robot", std::vector{pose});
  wpi::tunables::TunableRegistry::Update();

  CheckPose(field.GetRobotPose(), pose);
}

TEST_CASE_METHOD(Field2dTest,
                 "Field2dTest DashboardEditsObjectPosesThroughTunable",
                 "[wpilibc][smartdashboard]") {
  wpi::Field2d field;
  wpi::FieldObject2d* existing = field.GetObject("target");
  wpi::tunables::Publish("field", field);

  wpi::math::Pose2d first{1_m, 2_m, 30_deg};
  wpi::math::Pose2d second{3_m, 4_m, 60_deg};

  mock->SetStructVector<wpi::math::Pose2d>("/field/target",
                                           std::vector{first, second});
  wpi::tunables::TunableRegistry::Update();

  auto existingPoses = existing->GetPoses();
  REQUIRE(existingPoses.size() == 2u);
  CheckPose(existingPoses[0], first);
  CheckPose(existingPoses[1], second);
}

TEST_CASE_METHOD(Field2dTest,
                 "Field2dTest DashboardEditsObjectCreatedAfterTunablePublish",
                 "[wpilibc][smartdashboard]") {
  wpi::Field2d field;
  wpi::tunables::Publish("field", field);
  wpi::FieldObject2d* late = field.GetObject("target");

  wpi::math::Pose2d pose{1_m, 2_m, 30_deg};
  mock->SetStructVector<wpi::math::Pose2d>("/field/target", std::vector{pose});
  wpi::tunables::TunableRegistry::Update();

  auto latePoses = late->GetPoses();
  REQUIRE(latePoses.size() == 1u);
  CheckPose(latePoses[0], pose);
}

TEST_CASE_METHOD(Field2dTest,
                 "Field2dTest MoveAssignPublishedFromUnpublishedRepublishes"
                 "ObjectTunables",
                 "[wpilibc][smartdashboard]") {
  wpi::Field2d field;
  field.GetObject("stale");
  wpi::tunables::Publish("field", field);

  {
    wpi::Field2d replacement;
    wpi::math::Pose2d robotPose{1_m, 2_m, 30_deg};
    wpi::math::Pose2d targetPose{3_m, 4_m, 60_deg};
    replacement.SetRobotPose(robotPose);
    replacement.GetObject("target")->SetPose(targetPose);

    field = std::move(replacement);
  }

  CHECK_FALSE(mock->GetUid("/field/stale").has_value());

  auto robotPoses = mock->GetStructVector<wpi::math::Pose2d>("/field/Robot");
  REQUIRE(robotPoses.size() == 1u);
  CheckPose(robotPoses[0], {1_m, 2_m, 30_deg});

  auto targetPoses = mock->GetStructVector<wpi::math::Pose2d>("/field/target");
  REQUIRE(targetPoses.size() == 1u);
  CheckPose(targetPoses[0], {3_m, 4_m, 60_deg});

  wpi::math::Pose2d editedPose{5_m, 6_m, 90_deg};
  mock->SetStructVector<wpi::math::Pose2d>("/field/target",
                                           std::vector{editedPose});
  wpi::tunables::TunableRegistry::Update();

  auto movedTargetPoses = field.GetObject("target")->GetPoses();
  REQUIRE(movedTargetPoses.size() == 1u);
  CheckPose(movedTargetPoses[0], editedPose);
}
